### PR TITLE
test: add route with priority (E2E)

### PR DESCRIPTION
**Pull request by idbeta**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] E2E test
- [ ] New feature provided
- [ ] Improve performance

- Related issues
resolve https://github.com/apache/apisix-dashboard/issues/635

### CODE REVIEW

File: api/test/e2e/route_with_priority_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/server_port",
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+			ExpectBody:   "1982",
+			Sleep:        sleepTime,
+		},
+		{
+			caseDesc:     "delete route",
```
Comment (membphis) [inline api/test/e2e/route_with_priority_test.go:119]: Need a better desc

File: api/test/e2e/route_with_priority_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+			Object:       MangerApiExpect(t),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/routes/r1",
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime,
+		},
+		{
+			caseDesc:     "delete route",
```
Comment (membphis) [inline api/test/e2e/route_with_priority_test.go:128]: ditto

**Review by membphis — CHANGES_REQUESTED** [review]

**Review by membphis — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]
LGTM

File: api/test/e2e/route_with_priority_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+func TestRoute_with_priority(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "add route with priority (-1)",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+					  "uri": "/server_port",
```
Comment (nic-chen) [inline api/test/e2e/route_with_priority_test.go:32]: never use /server_port，unless you are testing multi upstream nodes.

Comment (membphis) [reply api/test/e2e/route_with_priority_test.go:32]: nice catch

Comment (juzhiyuan) [reply api/test/e2e/route_with_priority_test.go:32]: Could we add this tip to Testcase Guide?

Comment (membphis) [reply api/test/e2e/route_with_priority_test.go:32]: ping @idbeta

Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go:32]: this case is test the route priority, need the same uri to compare priority, in the current test docker container, `/server_port` is a more suitable uri.

Comment (nic-chen) [reply api/test/e2e/route_with_priority_test.go:32]: Sorry, I didn't see that this is the test priority, and it does need to be distinguished by port, but I suggest that multiple routes be created before verification, and the description is written more clearly.

File: api/test/e2e/route_with_priority_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+					  "methods": ["GET"],
+					  "priority": -1,
```
Comment (moonming) [inline api/test/e2e/route_with_priority_test.go:34]: do we support -1 as priority？

**Review by moonming — CHANGES_REQUESTED** [review]

**Comment by moonming** [comment]
> Please answer these questions before submitting a pull request
> 
> 
> 
> - Why submit this pull request?
> 
> - [x] E2E test
> 
> - [ ] New feature provided
> 
> - [ ] Improve performance
> 
> 
> 
> - Related issues
> 
> https://github.com/apache/apisix-dashboard/issues/635

Again, is this or will close #635?
Please learn from other PRs first

**Comment by idbeta** [comment]
> > Please answer these questions before submitting a pull request
> > 
> > * Why submit this pull request?
> > * [x]  E2E test
> > * [ ]  New feature provided
> > * [ ]  Improve performance
> > * Related issues
> > 
> > #635
> 
> Again, is this PR will close #635?
> Please learn from other PRs first

fixed. ping @moonming

File: api/test/e2e/route_with_priority_test.go (hunk 1 above)
Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go]: fixed

**Review by moonming — APPROVED** [review]

**Review by moonming — CHANGES_REQUESTED** [review]
do we support -1 as priority？

File: api/test/e2e/route_with_priority_test.go (hunk 4 above)
Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go:34]: yes, I have tested the priority of -2, -1, etc. and found that the priority supports negative values and conforms to the documentation, ` Larger value means higher priority. `.

Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go:34]: ping @moonming

Comment (moonming) [reply api/test/e2e/route_with_priority_test.go:34]: I think negative numbers are meaningless, maybe https://github.com/api7/lua-resty-radixtree should not allow negative numbers as a priority

File: api/test/e2e/route_with_priority_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+							  "port": 1980,
+							  "weight": 1
+						  }]
+					  }
+				  }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc:     "access the route to determine whether it meets the priority (pnly -1)",
```
[the remark this answers is not in the document]
Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go]: fixed.

File: api/test/e2e/route_with_priority_test.go (hunk 4 above)
Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go]: ok, I removed case which priority is a negative number. It is a `hidden attribute`.

**Comment by membphis** [comment]
> do we support -1 as priority？

Yes, we support now. 

here is the comparison step: https://github.com/api7/lua-resty-radixtree/blob/master/lib/resty/radixtree.lua#L179

**Comment by idbeta** [comment]
> Again, is this PR will close #635?
> Please learn from other PRs first

yes, ping @moonming

File: api/test/e2e/route_with_priority_test.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+func TestRoute_with_priority(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "add another route with no priority (default 0)",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+					  "uri": "/server_port",
```
Comment (moonming) [inline api/test/e2e/route_with_priority_test.go:32]: why use `/server_port`?

File: api/test/e2e/route_with_priority_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+			Method:       http.MethodGet,
+			Path:         "/server_port",
+			ExpectStatus: http.StatusOK,
+			ExpectBody:   "1981",
+			Sleep:        sleepTime,
+		},
+		{
+			caseDesc: "add another route with valid priority (1), upstream is different from the others",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
```
Comment (moonming) [inline api/test/e2e/route_with_priority_test.go:58]: we need to test the `get` method of manager API

Comment (moonming) [reply api/test/e2e/route_with_priority_test.go:58]: typo!!!! `MangerApiExpect`
what is `Manger`? @nic-chen @idbeta Please take it seriously！！！

**Review by moonming — CHANGES_REQUESTED** [review]

File: api/test/e2e/route_with_priority_test.go (hunk 6 above)
Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go:32]: When testing route priority, routes of different priorities need to use the same uri, and different routes return different content to determine whether the priority meets expectations, so I used the `/servert_port`.

File: api/test/e2e/route_with_priority_test.go (hunk 7 above)
Comment (idbeta) [reply api/test/e2e/route_with_priority_test.go:58]: fixed. ping @moonming Can you review it when you have time?

**Review by moonming — APPROVED** [review]

**Pull request merged**